src/Host/Client/Impl/BrokerServices/WebService.cs

```diff
@@ -15,71 +15,123 @@
+        private async Task RepeatUntilAuthenticatedAsync(Func<Task> action) {
+            await RepeatUntilAuthenticatedAsync(async () => {
+                await action();
+                return false;
+            });
         }
 
         public async Task<TResponse> HttpGetAsync<TResponse>(Uri uri, CancellationToken cancellationToken = default(CancellationToken)) {
-            var resonse = await HttpClient.GetAsync(uri, cancellationToken);
-            return JsonConvert.DeserializeObject<TResponse>(await resonse.Content.ReadAsStringAsync());
+            using (var response = await RepeatUntilAuthenticatedAsync(async () => EnsureSuccessStatusCode(await HttpClient.GetAsync(uri, cancellationToken)))) {
+                return JsonConvert.DeserializeObject<TResponse>(await response.Content.ReadAsStringAsync());
+            }
         }
 
         public Task<TResponse> HttpGetAsync<TResponse>(UriTemplate uriTemplate, CancellationToken cancellationToken = default(CancellationToken), params object[] args) =>
            HttpGetAsync<TResponse>(MakeUri(uriTemplate, args), cancellationToken);
 
         public async Task HttpPutAsync<TRequest>(Uri uri, TRequest request) {
             var requestBody = JsonConvert.SerializeObject(request);
-            var content = new StringContent(requestBody, Encoding.UTF8, "application/json");
-            EnsureSuccessStatusCode(await HttpClient.PutAsync(uri, content));
+
+            await RepeatUntilAuthenticatedAsync(async () => {
+                var content = new StringContent(requestBody, Encoding.UTF8, "application/json");
+                EnsureSuccessStatusCode(await HttpClient.PutAsync(uri, content)).Dispose();
+            });
         }
 
         public Task HttpPutAsync<TRequest>(UriTemplate uriTemplate, TRequest request, params object[] args) =>
             HttpPutAsync(MakeUri(uriTemplate, args), request);
 
         public async Task<TResponse> HttpPutAsync<TRequest, TResponse>(Uri uri, TRequest request, CancellationToken cancellationToken = default(CancellationToken)) {
             var requestBody = JsonConvert.SerializeObject(request);
-            var content = new StringContent(requestBody, Encoding.UTF8, "application/json");
-            var response = EnsureSuccessStatusCode(await HttpClient.PutAsync(uri, content, cancellationToken));
-            var responseBody = await response.Content.ReadAsStringAsync();
-            try {
-                return JsonConvert.DeserializeObject<TResponse>(responseBody);
-            } catch(JsonSerializationException ex) {
-                throw new ProtocolViolationException(ex.Message);
+
+            using (var response = await RepeatUntilAuthenticatedAsync(async () => {
+                var content = new StringContent(requestBody, Encoding.UTF8, "application/json");
+                return EnsureSuccessStatusCode(await HttpClient.PutAsync(uri, content, cancellationToken));
+            })) {
```

Comment:
Not sure what the comment refers to? `using` needs a brace here (and response needs to be disposed - we haven't actually been doing this properly before, which might account for some of memory leaks). 

I guess I could declare and initialize `response` outside of `using`, and then just do `using (response) { ... }`. Would this be preferable readability-wise?